### ISSUE TITLE
Handling outputs from the same model with different likelihoods / score functions

### ISSUE THREAD

**Issue by ben18785** [pints-team/pints#1403]
PKPD models will likely need outputs from the same model with different likelihoods. E.g. one output follows a normal model; another follows a log-normal. At the moment, this sort of thing is not possible in PINTS without lots of manual hacking. Similarly, it is not possible to natively handle an inference problem with multiple outputs where those outputs are measured at different times. 

I wanted to open the forum on this. Here's a potential proposal.

We create a `MasterProblemCollection` with the following methods:

- initialised with a forward model and values as is usual. It is also (crucially) initialised with some way of indexing which outputs correspond to which `SubProblem` (see below). At a simple level, this could be a list `[1,1,1,2,2,2,2,3,3,3,3,3]` which would indicate that the first three forward model outputs correspond to subproblem1, the second four to subproblem2 and the last five to subproblem3. It is also initiated with a list of `times` lists: one list of times for each `SubProblem`.
- `evaluate` takes an argument `index` (in the above example, this would be either 1, 2 or 3) which specifies which output set to return. The first time evaluate is called, the result is cached so that the model is only solved once -- not once for each output set.
- `n_outputs` also takes `index` and returns the relevant number of outputs for the `SubProblem`
- `times` takes `index` and returns the appropriate time list

We also create a `SubOutputProblem`:

- intialised with a `MasterProblemCollection` and an `index`
- `evaluate` calls `MasterProblemCollection.evaluate(index)`
- and so on for `n_outputs` etc

## Example use case
User has a forward model with three outputs and would to model the first two using a `GaussianLogLikelihood` and the third with `LogNormalLogLikelihood`.

They would do:

```
times_1 = [1, 2, 4, 6]
times_2 = [3, 5]
times = [times_1, times_2]
index_list = [1, 1, 2]
master = MasterProblemCollection(forward_model, times, value, index_list)

subproblem_1 = SubProblem(master, 1)
subproblem_2 = SubProblem(master, 2)

loglikelihood_1 = GaussianLogLikelihood(subproblem_1)
loglikelihood_2 = LogNormalLogLikelihood(subproblem_2)
logprior_1 = ...
logprior_2 = ...

logposterior_1 = LogPosterior(loglikelihood_1, logprior_1)
logposterior_2 = LogPosterior(loglikelihood_2, logprior_2)
```

The user could then create a wrapper to create a `LogPosterior` class that just returns the sum of `logposterior_1` and `logposterior_2` which they could use for inference.

@martinjrobins @MichaelClerx @chonlei (and anyone else!) interested to hear thoughts on this

**Comment by martinjrobins** [comment]
just a note to say we will definitely need this for the pkpdapp inference. Different outputs of the ODE model will have different noise models, hence the need for different distributions on different outputs. The different outputs might also be measured at different times, e.g. if you are measuring the concentration of the drug as well as the size of a tumour, these wouldn't neccessarily be at the same time.

**Comment by ben18785** [comment]
Yep, spot on with what I was thinking re: 1 and, yes, happy to shy away from "master"!

Re: filters, I'm not sure I get ya? Also, currently an issue with MultiSeriesProblem is that this doesn't allow different time arrays for each output, so I guess that'd need to be re-engineered?

**Comment by MichaelClerx** [comment]
Basically, in your proposal you start by creating a superproblem that then needs to combine some sub problems, run a sim, and split its output again.
I'm wondering why you don't cut out the first step, and just create methods to split a problem's output into sub outputs?

So you'd have an ordinary MultiSeriesProblem, but then you'd make sub problems based on its output, and finally combine those sub problems into a single likelihood (although you could hide output as well, using this method)

**Comment by MichaelClerx** [comment]
Oooh that's a good point, I was thinking about the simulation much more than the data 😅 

Ideally I guess you'd have a ``None`` in the data set to denote no value, but may need to use a ``nan`` so that it fits into a numpy doubles array...

But if we start allowing that, then all the likelihoods etc. would need to do some clever filtering? Or... do we write that ``nan`` may appear in a multioutputseries problem's values, but that we expect the user to construct a clever score/likelihood so that this isn't a problem?

**Comment by MichaelClerx** [comment]
> So would the MultiSeriesProblem take all the times where all outputs were evaluated?

Yup!

**Comment by MichaelClerx** [comment]
How were you planning to handle data in the super/sub problem set-up?

**Comment by ben18785** [comment]
So the way I'd handle data would be the same way as for times:

```
## times and values for first output
times_1 = [1, 2, 4, 6]
values_1 = [2.4, 3.4, 5.5, 5.5]

## times and values for second output
times_2 = [3, 5]
values_2 = [-0.3, 0.5]

## combine all
times_all = [times_1, times_2]
values_all = [values_1, values_2]
index_list = [1, 1, 2]
master = MasterProblemCollection(forward_model, times_all, values_all, index_list)
```

So, we could put `nan`s in the numpy arrays, although this feels a little clunky. I do wonder though whether this filtering means users are more likely to make mistakes -- specifically if their filters don't grab the right values?

**Comment by MichaelClerx** [comment]
Is `index_list` correct in that example?

**Comment by MichaelClerx** [comment]
```
model = MyNiceModel()
model_with_fixed_times = FixedTimesModel(model, times)
submodel1 = FilteredModel(model_with_fixed_times, times, outputs) --> can check its times are a subset of parent model's times
submodel2 = FilteredModel(...)
```

yuck!
